File: samples/cpp/automatic_speech_recognition/asr_forced_aligner.cpp
// Copyright (C) 2026 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

#include <filesystem>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>

#include "audio_utils.hpp"
#include "openvino/genai/automatic_speech_recognition/forced_aligner.hpp"

int main(int argc, char* argv[]) try {
    if (argc < 4 || argc > 6) {
        throw std::runtime_error(std::string{"Usage: "} + argv[0] +
                                 " <MODEL_DIR> \"<WAV_FILE_PATH>\" \"<TRANSCRIPT>\" [DEVICE] [LANGUAGE]");
    }

    std::filesystem::path models_path = argv[1];
    std::string wav_file_path = argv[2];
    std::string transcript = argv[3];
    std::string device = (argc >= 5) ? argv[4] : "CPU";
    std::string language = (argc >= 6) ? argv[5] : "english";

    ov::genai::ASRForcedAligner aligner(models_path, device);

    // Forced aligner expects normalized audio with a sample rate of 16kHz
    ov::genai::RawSpeechInput raw_speech = utils::audio::read_wav(wav_file_path);
    auto words = aligner.align(raw_speech, transcript, ov::genai::language(language));

    std::cout << std::fixed << std::setprecision(2);
    for (const auto& word : words) {
        std::cout << "[" << word.start_ts << ", " << word.end_ts << "]: " << word.text << "\n";
    }

} catch (const std::exception& error) {
    try {
        std::cerr << error.what() << '\n';
    } catch (const std::ios_base::failure&) {
    }
    return EXIT_FAILURE;
} catch (...) {
    try {
        std::cerr << "Non-exception object thrown\n";
    } catch (const std::ios_base::failure&) {
    }
    return EXIT_FAILURE;
}

File: samples/python/automatic_speech_recognition/asr_forced_aligner.py
#!/usr/bin/env python3
# Copyright (C) 2026 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import argparse
import openvino_genai
import librosa


def read_wav(filepath):
    raw_speech, _ = librosa.load(filepath, sr=16000)
    return raw_speech.tolist()


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("model_dir", help="Path to the forced-aligner model directory")
    parser.add_argument("wav_file_path", help="Path to the WAV file")
    parser.add_argument("transcript", help="Transcript to align to the audio")
    parser.add_argument("device", nargs="?", default="CPU", help="Device to run the model on (default: CPU)")
    parser.add_argument("--language", default="english", help="Transcript language (default: english)")
    args = parser.parse_args()

    aligner = openvino_genai.ASRForcedAligner(args.model_dir, args.device)

    raw_speech = read_wav(args.wav_file_path)
    words = aligner.align(raw_speech, args.transcript, language=args.language)

    for word in words:
        print(f"[{word.start_ts:.2f}, {word.end_ts:.2f}]: {word.text}")


if "__main__" == __name__:
    main()
